user/ntfs-3g/src/ntfs-3g.8
.\" Copyright (c) 2005-2006 Yura Pakhuchiy.
.\" Copyright (c) 2005 Richard Russon.
.\" Copyright (c) 2006-2008 Szabolcs Szakacsits.
.\" This file may be copied under the terms of the GNU Public License.
.\"
.TH NTFS-3G 8 "April 2008" "ntfs-3g 1.2531"
.SH NAME
ntfs-3g \- Third Generation Read/Write NTFS Driver
.SH SYNOPSIS
.B ntfs-3g
.I volume mount_point
\fB[-o \fIoption\fP\fB[,...]]\fR
.br
.B mount \-t ntfs-3g
.I volume mount_point
\fB[-o \fIoption\fP\fB[,...]]\fR
.SH DESCRIPTION
\fBntfs-3g\fR is an NTFS driver, which can
create, remove, rename, move files, directories, hard links, and
streams; it can read and write files, including 
streams and sparse files; it can handle special files like 
symbolic links, devices, and FIFOs; moreover it can also read
transparently compressed files.
.PP
The \fIvolume\fR to be mounted can be either a block device or 
an image file.
.SS Access Handling and Security
By default, files and directories are owned by the effective 
user and group of the mounting process and everybody has
full read, write, execution and directory browsing permissions.
If you want to use permissions handling then use the
.B uid
and/or the
.B gid 
options together with the 
.B umask,
or
.B fmask
and
.B dmask
options.
.PP
Windows users have full access to the files created by 
.B ntfs-3g.
.PP
If 
.B ntfs-3g 
is set setuid-root then non-root users will 
be also able to mount volumes.
.SS Windows Filename Compatibility
NTFS supports several filename namespaces: DOS, Win32 and POSIX. While the
\fBntfs-3g\fR driver handles all of them, it always creates new files in the 
POSIX namespace for maximum portability and interoperability reasons. 
This means that filenames are case sensitive and all characters are
allowed except '/' and '\\0'. This is perfectly legal on Windows, though
some application may get confused. If you find so then please report it
to the developer of the relevant Windows software.
.SS Alternate Data Streams (ADS)
NTFS stores all data in streams. Every file has exactly one unnamed
data stream and can have many named data streams.  The size of a file is the
size of its unnamed data stream.  By default, \fBntfs-3g\fR will only read
the unnamed data stream.
.PP
By using the options "streams_interface=windows", you will be able to read
any named data streams, simply by specifying the stream's name after a colon.
For example:
.RS
.sp
cat some.mp3:artist
.sp
.RE
Named data streams act like normals files, so you can read from them, write to
them and even delete them (using rm).  You can list all the named data streams
a file has by getting the "ntfs.streams.list" extended attribute.
.SH OPTIONS
Below is a summary of the options that \fBntfs-3g\fR accepts.
.TP
\fBuid=\fP\fIvalue\fP and \fBgid=\fP\fIvalue\fP
Set the owner and the group of files and directories. The values are numerical.
The defaults are the uid and gid of the current process.
.TP
.BI umask= value
Set the  bitmask of the file and directory permissions that are not
present. The value is given in octal. The default value is 0 which
means full access to everybody.
.TP
.BI fmask= value
Set the  bitmask of the file permissions that are not present. 
The value is given in octal. The default value is 0 which
means full access to everybody.
.TP
.BI dmask= value
Set the  bitmask of the directory permissions that are not
present. The value is given in octal. The default value is 0 which
means full access to everybody.
.TP
.B ro
Mount filesystem read\-only. Useful if Windows is hibernated or the
NTFS journal file is unclean.
.TP
.BI locale= value
This option can be useful if your language specific locale environment 
variables are not set correctly or at all in your operating system.
In such cases, the national characters can be made visible by using this 
option. Please see more information about this topic at
http://ntfs-3g.org/support.html#locale
.TP
.B force
Force the mounting even if the NTFS logfile is unclean. The logfile 
will be unconditionally cleared. Use this option with caution and for 
your own responsibility.
.TP
.B remove_hiberfile
Unlike in case of read-only mount, the read-write mount is denied if 
the NTFS volume is hibernated. One needs either to resume Windows and
shutdown it properly, or use this option which will remove the Windows
hibernation file. Please note, this means that the saved Windows 
session will be completely lost. Use this option for your own 
responsibility.
.TP
.B atime, noatime, relatime
The 
.B atime 
option updates inode access time for each access.

The 
.B noatime 
option disables inode access time updates which can speed up
file operations and prevent sleeping (notebook) disks spinning 
up too often thus saving energy and disk lifetime.

The
.B relatime 
option is very similar to 
.B noatime. 
It  updates inode access times relative to modify or change time. 
The access time is only updated if the previous access time was earlier 
than the current modify or change time. Unlike
.B noatime
this option doesn't break applications that need to know 
if a file has been read since the last time it was modified.
This is the default behaviour.
.TP
.B show_sys_files
Show the system files in directory listings. 
Otherwise the default behaviour is to hide the system files.
Please note that even when this option is specified, "$MFT"
may not be visible due to a glibc bug.
Furthermore, irrespectively of show_sys_files, all
files are accessible by name, for example you can always do
"ls \-l '$UpCase'".
.TP
.BI max_read= value
With this option the maximum size of read operations can be set.
The default is infinite.  Note that the size of read requests is
limited anyway to 32 pages (which is 128kbyte on i386).
.TP
.B silent
Do nothing on chmod and chown operations, but do not return error. 
This option is on by default.
.TP
.B no_def_opts
By default ntfs-3g acts as "silent" was passed to it,
this option cancel this behaviour.
.TP
.BI streams_interface= value
This option controls how the user can access Alternate Data Streams (ADS)
or in other words, named data streams. It can be set
to, one of \fBnone\fR, \fBwindows\fR or \fBxattr\fR. If the option is set to
\fBnone\fR, the user will have no access to the named data streams.  If it's set
to \fBwindows\fR, then the user can access them just like in Windows (eg. cat
file:stream). If it's set to \fBxattr\fR, then the named data streams are
mapped to xattrs and user can manipulate them using \fB{get,set}fattr\fR
utilities. The default is \fBnone\fR.
.TP
.B debug
Makes ntfs-3g to not detach from terminal and print a lot of debug output from
libntfs-3g and FUSE.
.TP
.B no_detach
Same as above but with less debug output.
.SH EXAMPLES
Mount /dev/sda1 to /mnt/windows:
.RS
.sp
.B ntfs-3g /dev/sda1 /mnt/windows
.sp
.RE
or
.RS
.sp
.B mount -t ntfs-3g /dev/sda1 /mnt/windows
.sp
.RE
Read\-only mount /dev/sda5 to /home/user/mnt and make user with uid 1000 
to be the owner of all files:
.RS
.sp
.B ntfs-3g /dev/sda5 /home/user/mnt \-o ro,uid=1000
.sp
.RE
/etc/fstab entry for the above:
.RS
.sp
.B /dev/sda5 /home/user/mnt ntfs\-3g ro,uid=1000 0 0
.sp
.RE
Unmount /mnt/windows:
.RS
.sp
.B umount /mnt/windows
.sp
.RE
.SH EXIT CODES
To facilitate the use of the
.B ntfs-3g
driver in scripts, an exit code is returned to give an indication of the 
mountability status of a volume. Value 0 means success, and all other
ones mean an error. The unique error codes are documented in the
.BR ntfs-3g.probe (8)
manual page.
.SH KNOWN ISSUES
Please see 
.RS
.sp
http://ntfs-3g.org/support.html
.sp
.RE
for common questions and known issues.
If you would find a new one in the latest release of
the software then please send an email describing it
in detail. You can contact the 
development team on the ntfs\-3g\-devel@lists.sf.net
address.
.SH AUTHORS
.B ntfs-3g 
was based on and a major improvement to ntfsmount and libntfs which were
written by Yura Pakhuchiy and the Linux-NTFS team. The improvements were 
made, the ntfs-3g project was initiated and currently led by long time 
Linux-NTFS team developer Szabolcs Szakacsits (szaka@ntfs-3g.org).
.SH THANKS
Several people made heroic efforts, often over five or more
years which resulted the ntfs-3g driver. Most importantly they are 
Anton Altaparmakov, Richard Russon, Szabolcs Szakacsits, Yura Pakhuchiy,
Yuval Fledel, and the author of the groundbreaking FUSE filesystem development 
framework, Miklos Szeredi.
.SH SEE ALSO
.BR ntfs-3g.probe (8),
.BR ntfsprogs (8),
.BR attr (5),
.BR getfattr (1)


History
<b>Changes:</b><br>
<b>1.0.x</b><br>
1) Fix configs scripts for support bash >=4 and change product name.<br>
2) Kernel version up to 2.6.21.7.<br>
3) Redesign filesystem. Busybox up to 1.15.3.<br>
4) PPPD up to 2.4.5 with more patches as acell-ptpp/l2tp/gigaworld/allow-mppe/etc.<br>
5) Add ACELL-PPTP Kernel support.<br>
6) Mount ramfs particions in kernel mode.<br>
7) BRIDGE: Pause frames shouldn't be passed up by driver anyway.<br>
8) Add ESFQ support.<br>
9) Initial auto detect memory size (fix me 32bit mode support).<br>
10) Add udpxy.<br>
11) Code cleanup.<br>
12) Remove ASUS Code.<br>
13) Original Ralink gui now work.<br>
14) pptp/l2tp/pppoe tun now works.<br>
15) Full rc scrips is write in rwfs.<br>
16) RT3052F embed switch full work now.<br>
17) Rewrite pptp/l2tp/pppoe scripts to support buggy nas (corbina etc).<br>
18) miniupnpd now work too.<br>
19) Add rwfs support (fs save | fs restore | fs load).<br>
20) Auto set mss size.<br>
21) Masqrade now work correctly.<br>
22) IMQ support added.<br>
23) Initial shaper scripts (default priomap and esfq for decrase torrents prio).<br>
24) Kernel ip account support added.<br>
25) Fix crash on sta list.<br>
26) Update userspace daemons and utility.<br>
27) Some busybox fixes.<br>
28) Auto generate UUID for UPNP service on first upnpd start.<br>
29) Check if ppp is up do not change ip from dhcp.<br>
30) Regenerate mac if drop to defaults.<br>
31) Some services start only then start goahead.<br>
32) Replace some code from original Ralink SDK version 3.3.<br>
33) Integrate new web gui in current sources tree.<br>
34) Static routes get from dhcp server (options 79).<br>
35) busybox update to 1.16.0 version.<br>
36) Replace ntpclient to ntpd from new busybox.<br>
37) Add multiple default gateway support for load balacing.<br>
38) Fix boot on 4/16 devices.<br>
39) RootFS on flash work OK.<br>
40) Fix resolve server ip for pptp/l2tp.<br>
41) Add support others type falsh.<br>
42) Job control fix.<br>
43) Fix some daemons crash with error build in future.<br>
44) fs save now save current date and time to rwfs.<br>
45) Fix configure esw mode and more modes add.<br>
46) Fix pppoe reconnect.<br>
47) Fix pppoe long discovery.<br>
48) Restore default gateway after VPN down.<br>
49) WPS wscd daemon now work.<br>
50) Fix dhcp lease parse.<br>
51) busybox update to 1.16.1 version.<br>
52) Fix pptp/pppoe/l2tp redials.<br>
53) Radvd up to 1.6 version.<br>
54) Masqrade for vpn move to ip-up/ip-down scripts.<br>
55) NAT to ISP local net parallel then vpn work now.<br>
56) Never increase MSS, even when setting it.<br>
57) Backport small bridge patch.<br>
58) Backport sock memalloc skb size from 2.6.23.4 kernel.<br>
59) ACCEL-PPTP add module param description and mtu parametr.<br>
60) Patch to fix cut end symbol at hostname in udhcpd lease.<br>
61) Backport jhash3 patch from upstream.<br>
62) More ipv6 fixes.<br>
63) Improve pppoe perf.<br>
64) pptp module patches from wl500 project.<br>
65) More user/password length to pptp/l2tp mode.<br>
66) More clean init.d scripts.<br>
67) Fix kernel crash on recive multicast.<br>
68) Autodetect RAM size (16-64) work now for all devices.<br>
69) UDPXY work now.<br>
70) Igmpproxy work now.<br>
71) Use TMPFS instead RAMFS. Save memory and increase speed.<br>
72) Fix restore nvram after flash at new device.<br>
73) LZMA crash fix.<br>
74) Mtdblock crash and memory leak fix.<br>
75) Remove cifs. User ssh scp instead.<br>
76) Update wifi driver to 2.3.0.0 version from Ralink.<br>
77) More fix build enviroment.<br>
78) Bridge code fixup and more fast.<br>
<br>
<b>1.1.x</b><br>
1) New WebUI based on goahead.<br>
2) New wifi driver, correct work with broadcom wifi based adapters, WMM, etc...<br>
3) Set of corrections in a kernel (netfiler/routing/ipv6).<br>
4) Rewrite logic in init.d for speedup.<br>
5) Add simple software conntrack fastpath.<br>
6) Cleannat module support for clear conntrack table.<br>
7) More connrack helpers and protocol support.<br>
8) Kernel backport fixes from stable tree.<br>
9) Busybox update to 1.17.x tree.<br>
10) Some scrits rewrite for speedup.<br>
11) Remove hldc and vj support for pppd.<br>
12) ppp connection optimization in kernel.<br>
13) DMZ must works.<br>
14) Sync from 3.4.0.1 RalinkSDK from trandnet TEW-691GR GPL code.<br>
15) libnvram fix crash on reliad settings and some leaks.<br>
16) Fix freeze in 802.11n mode change in EthernetConverterMode.<br>
17) ppp use metric 10 for dgw replace.<br>
18) add more workaround for buggy nas.<br>
19) Portforward work now.<br>
20) MAC/IP/PORT filterd work now.<br>
21) pptp/l2tp now work with new corbina bras.<br>
22) Get staticroutes and msstatic routes from dhcp server work now.<br>
23) WebUI Add watchdog enable/disable function.<br>
24) Fix reboot at apply change in ethernet converter mode.<br>
25) Fix lldt control in webui.<br>
26) Add interface select for macip filter and port forward in WebUI.<br>
27) Add 802.11g/n mixed mode and set default.<br>
28) Add samba for smb and wins services.<br>
29) portforward and ipport filter scripts now reboot at start.<br>
30) Content filters based on webstr now works.<br>
31) Use WLAN_MAC_ADDR, WAN_MAC_ADDR, LAN_MAC_ADDR variable in nvram for set macs to interfaces.<br>
32) STA profiles now works.<br>
33) Add netmask for ip filter pages.<br>
34) Add asus rt-g32 full support.<br>
35) Fix part table in SPI and nand flash types.<br>
36) pppd up to current git src 24 Aug 2010.<br>
37) xl2tpd up to 1.2.7 version.<br>
38) Fix conntrack flush.<br>
39) Wireless pages split and optimize.<br>
40) VPN page add auth type select.<br>
41) Fix trunclate nvram values at aplly change and device mode select.<br>
42) Fix restart nvram_daemon.<br>
43) Kernel fix lock in mtdchar, more small fixes and cleanups.<br>
44) Fix keep mac adresses at push reset button.<br>
45) libnvram use mtdblock instead mtdchar for write/read function.<br>
46) Fix unlocked kernel write to mtd. Use kernel_lock for fix concurrent write issue.<br>

user/rt2880_app/ntfs-3g/src/ntfs-3g.probe.c
/*
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 * MA 02111-1307 USA
 */
/**
 * ntfs-3g.probe - Probe NTFS volume mountability
 *
 * Copyright (c) 2007-2009 Szabolcs Szakacsits
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program (in the main directory of the NTFS-3G
 * distribution in the file COPYING); if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "config.h"

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#include <getopt.h>

#include "compat.h"
#include "volume.h"
#include "misc.h"

typedef enum {
	PROBE_UNSET,
	PROBE_READONLY,
	PROBE_READWRITE
} probe_t;

static struct options {
	probe_t  probetype;
	char	 *device;
} opts;

static const char *EXEC_NAME = "ntfs-3g.probe";

static const char *usage_msg = 
"\n"
"%s %s - Probe NTFS volume mountability\n"
"\n"
"Copyright (C) 2007 Szabolcs Szakacsits\n"
"\n"
"Usage:    %s <--readonly|--readwrite> <device|image_file>\n"
"\n"
"Example:  ntfs-3g.probe --readwrite /dev/sda1\n"
"\n"
"%s";

static int ntfs_open(const char *device)
{
	ntfs_volume *vol;
	unsigned long flags = 0;
	int ret = NTFS_VOLUME_OK;
	
	if (opts.probetype == PROBE_READONLY)
		flags |= MS_RDONLY;

	vol = ntfs_mount(device, flags);
	if (!vol)
		ret = ntfs_volume_error(errno);

	ntfs_umount(vol, FALSE);

	return ret;
}

static void usage(void)
{
	ntfs_log_info(usage_msg, EXEC_NAME, VERSION, EXEC_NAME, ntfs_home);
}

static int parse_options(int argc, char *argv[])
{
	int c;

	static const char *sopt = "-hrw";
	static const struct option lopt[] = {
		{ "readonly",	no_argument,	NULL, 'r' },
		{ "readwrite",	no_argument,	NULL, 'w' },
		{ "help",	no_argument,	NULL, 'h' },
		{ NULL,		0,		NULL,  0  }
	};

	opterr = 0; /* We handle errors. */
	opts.probetype = PROBE_UNSET;

	while ((c = getopt_long(argc, argv, sopt, lopt, NULL)) != -1) {
		switch (c) {
		case 1:	/* A non-option argument */
			if (!opts.device) {
				opts.device = ntfs_malloc(PATH_MAX + 1);
				if (!opts.device)
					return -1;
				
				strncpy(opts.device, optarg, PATH_MAX);
				opts.device[PATH_MAX] = 0;
			} else {
				ntfs_log_error("%s: You must specify exactly "
					       "one device\n", EXEC_NAME);
				return -1;
			}
			break;
		case 'h':
			usage();
			exit(0);
		case 'r':
			opts.probetype = PROBE_READONLY;
			break;
		case 'w':
			opts.probetype = PROBE_READWRITE;
			break;
		default:
			ntfs_log_error("%s: Unknown option '%s'.\n", EXEC_NAME,
				       argv[optind - 1]);
			return -1;
		}
	}

	if (!opts.device) {
		ntfs_log_error("ERROR: %s: Device is missing\n", EXEC_NAME);
		return -1;
	}

	if (opts.probetype == PROBE_UNSET) {
		ntfs_log_error("ERROR: %s: Probe type is missing\n", EXEC_NAME);
		return -1;
	}

	return 0;
}

int main(int argc, char *argv[])
{
	int err;

	ntfs_log_set_handler(ntfs_log_handler_stderr);

	if (parse_options(argc, argv)) {
		usage();
		exit(NTFS_VOLUME_SYNTAX_ERROR);
	}

	err = ntfs_open(opts.device);

	free(opts.device);
	exit(err);
}

